Update OAuth Specifications and add Prompt parameter documentation (#489)
bluesky-social/atproto-website: src/app/[locale]/specs/oauth/en.mdx

## Clients

This section describes requirements for OAuth clients, which are enforced by Authorization Servers.

OAuth client software is identified by a globally unique `client_id`. Distinct variants of client software may have distinct `client_id` values; for example the browser app and Android (mobile OS) variants of the same software might have different `client_id` values. As required by the [`draft-ietf-oauth-client-id-metadata-document`](https://datatracker.ietf.org/doc/draft-ietf-oauth-client-id-metadata-document/) specification draft, the `client_id` must be a fully-qualified web URL from which the client-metadata JSON document can be fetched. For example, `https://app.example.com/oauth-client-metadata.json`. Some more about the `client_id`:

- it must be a well-formed URL, following the W3C URL specification
- the schema must be `https://`, and there must not be a port number, username or password included.

> - the path does not need to include `oauth-client-metadata.json`, but it is helpful convention.

Author: Debatable here as to whether we want to specify that the URL should not include query string component. https://www.ietf.org/archive/id/draft-ietf-oauth-client-id-metadata-document-00.html#section-3

Author: @matthieusieben how do you feel about prohibiting querystrings in the CIMD URLs?

Authorization Servers which support both the atproto OAuth profile and other forms of OAuth should take care to prevent `client_id` value collisions. For example, `client_id` values for clients which are not auto-registered should never have the prefix `https://` or `http://`.

<Note>

> There is a special exception for `client_id` values starting with `http://localhost`, which MAY include a port number, see the [Localhost Client Development](#localhost-client-development) section for details.

Reviewer: IIRC the `http://localhost` client ids must not contain a port number in the "host" part of the client id uri. They may specify query parameters though (most notably for redirect uris & scopes)

### Localhost Client Development

<Note>
Localhost Client IDs are not a standard specification, and only provide access as a public client. The Client ID Metadata Documents internet draft recommends using a CIMD Service instead.

> </Note>

Reviewer: We actually have other ideas here.

> </Note>

Author: definitely interested to hear them!

Native clients are also allowed to use an HTTPS URL. In this case, the URL origin must be the same as the `client_id`. One example use-case is ["Apple Universal Links"](https://developer.apple.com/documentation/xcode/allowing-apps-and-websites-to-link-to-your-content).

Clients may include additional optional authorization request parameters - and servers may process them - but they are not required to. Refer to other OAuth standards and the [IANA OAuth parameter registry](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml).

### Prompt Parameter

Authorization Servers may support various `prompt` parameter values. These values are advertised by the `prompt_values_supported` property in the [Authorization Server Metadata](#server-metadata). This is a parameter that is borrowed from the OpenID Connect 1.0 family of specifications: [OpenID Connect Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) and [Initiating User Registration via OpenID Connect 1.0](https://openid.net/specs/openid-connect-prompt-create-1_0.html).

> <Note>
> The OpenID Connect Core 1.0 specification allows for multiple `prompt` parameter values in the Authorization Request, however the reference implementation only supports a singular value. As such, using multiple `prompt` parameters in one Authorization Request is NOT RECOMMENDED.
> </Note>

Reviewer: If we support this spec, we should make the reference implementation fully compatible.

```suggestion
```

Author: Currently I don't think we do, so we'd need to modify the oauth-provider to support multiple.

For understanding how the `prompt` parameter affects the Authorization Server User Interface, see the ["Authorization Interface"](#authorization-interface) section.

> Currently known values for the `prompt` parameter include:

Author: ```suggestion
Current known values for the `prompt` parameter include:
```

- `none`: The Authorization Server MUST NOT display any authentication or consent user interface pages. An error is returned if an End-User is not already authenticated or the Client does not have pre-configured consent for the requested authorization request or does not fulfill other conditions for processing the request. The error code will typically be `login_required`, `interaction_required`, or [another error code](https://openid.net/specs/openid-connect-core-1_0.html#AuthError). This can be used as a method to check for existing authentication and/or consent.

> - `consent`: The Authorization Server SHOULD prompt the End-User for consent before returning information to the Client. If it cannot obtain consent, it MUST return an error, typically `consent_required` or `access_denied`.

Author: Is there actually a case where you'll get back `consent_required` as an error though? I don't think that's implemented.

`access_denied` is really kinda separate, and we should probably document that up in a section on "Handling redirect URI requests" or something, and say that you may receive back the `error` parameter, not just a `code` parameter.

> - `consent`: The Authorization Server SHOULD prompt the End-User for consent before returning information to the Client. If it cannot obtain consent, it MUST return an error, typically `consent_required` or `access_denied`.

Reviewer: This list only contains values for the oauth RFC. See below:

https://github.com/bluesky-social/atproto/blob/324727933699bee63a1db3b1fba040e067511ad4/packages/oauth/oauth-provider/src/errors/consent-required-error.ts#L4-L12

> - `consent`: The Authorization Server SHOULD prompt the End-User for consent before returning information to the Client. If it cannot obtain consent, it MUST return an error, typically `consent_required` or `access_denied`.

Author: Which oauth RFC? I'd based this on the OIDC prompt parameter as wasn't aware of an oauth RFC? https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

> - `login`: The Authorization Server SHOULD prompt the End-User for reauthentication. If it cannot reauthenticate the End-User, it MUST return an error, typically `login_required`.

Author: I'm not sure if this error behaviour is actually implemented in the reference server.

Reviewer: https://github.com/bluesky-social/atproto/blob/324727933699bee63a1db3b1fba040e067511ad4/packages/oauth/oauth-provider/src/oauth-provider.ts#L710

Reviewer: `login_required` is for failed sso. If the user is presented with the UI and refuses to authenticate, it is an explicit deny

```suggestion
- `login`: The Authorization Server SHOULD prompt the End-User for reauthentication. If it cannot reauthenticate the End-User, it MUST return an error, typically `access_denied`.
```

Author: This language for `login_required` was taken from the [OpenID Connect spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest):
>  The Authorization Server SHOULD prompt the End-User for reauthentication. If it cannot reauthenticate the End-User, it MUST return an error, typically login_required.

> - `select_account`: The Authorization Server SHOULD prompt the End-User to select a user account. This enables an End-User who has multiple accounts at the Authorization Server to select amongst the multiple accounts that they might have current sessions for. If it cannot obtain an account selection choice made by the End-User, it MUST return an error, typically `account_selection_required`. 

Author: Matthieu just added this to the reference implementation.

> - `create`: The Authorization Server SHOULD prompt the End-User for account creation rather than the login. Upon successfully creating an account, the user should be shown the consent screen. Support for this parameter was only recently added. See the ["Initiating User Registration"](#initiating-user-registration) section below.

Reviewer: ```suggestion
- `create`: The Authorization Server SHOULD prompt the End-User for account creation rather than the login. Upon successfully creating an account, the user should be shown the consent screen. See the ["Initiating User Registration"](#initiating-user-registration) section below.
```

> Authorization Servers SHOULD ignore the `prompt` parameter or its value if they do not support this user experience.

Reviewer: I think we should aim to for a strong spec, with a few choices as possible for implementations: This makes interop simpler between services, making implementing oauth client/as simpler.

So I would change this with something like

```suggestion
AS MUST support the `prompt` parameter, and MUST advertise the supported values in `prompt_values_supported`. Clients MAY account for AS not up-to-date with this spec by detecting missing support for the `prompt` parameter (by detecting the missing `prompt_values_supported`.
```

## Authorization Servers

To enable browser apps, Authorization Servers must support [HTTP CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) requests/headers on relevant endpoints, including server metadata, auth requests (PAR), and token requests.

### Server Metadata

> Both Resource Servers (PDS instances) and Authorization Servers (PDS or [PDS Entryway](https://docs.bsky.app/docs/advanced-guides/entryway)) need to publish metadata files at well-known HTTPS endpoints.

Author: @axfelix perhaps we do:
```suggestion
In AT Protocol, we have both Resource Servers and Authorization Servers, which may be combined into one service or as two separate services. When running the Authorization Server separately from the PDS, we end up with the PDS being the resource server and the Entryway being the Authorization Server.

To allow clients to discover how to interact with the Authorization Server, both the Resource Server and Authorization Server need to publish metadata files at well-known HTTPS endpoints.
```

- `require_pushed_authorization_requests` (boolean): must be `true`. See ["PAR"](#pushed-authorization-requests-par) section.
- `pushed_authorization_request_endpoint` (string, required): must be the PAR endpoint URL. See ["PAR"](#pushed-authorization-requests-par) section.
- `dpop_signing_alg_values_supported` (array of strings, required): currently must include `ES256`. See ["DPoP"](#demonstrating-proof-of-possession-dpop) section.
- `require_request_uri_registration` (boolean, optional): default is `true`; does not need to be set explicitly, but must not be `false`
- `client_id_metadata_document_supported` (boolean, required): must be `true`. See ["Client ID Metadata"](#client-id-metadata-document) section.
- `prompt_values_supported` (array of strings, optional but recommended): The `prompt` parameter values supported. May include `create` for initiating registration. See the ["Prompt Parameter"](#prompt-parameter) and ["Initiating User Registration"](#initiating-user-registration) sections.

The `issuer` ("origin") is the overall identifier for the Authorization Server.

### Authorization Interface

> The Authorization Server (PDS or [PDS Entryway](https://docs.bsky.app/docs/advanced-guides/entryway)) must implement a web interface for users to authenticate with the server, approve (or reject) Authorization Requests from clients, and manage active sessions. This is called the "Authorization Interface".

Reviewer 2: The entryway is in quite a small category of "complicated implementations that are worth documenting but are relatively Bluesky-specific." After this recent rewrite to the ATProto docs the number of times we link out to the Bluesky Docs will be reduced from something like 20 to something like 2, but this is one of them for now.

Probably we should port this page into the new auth section and rewrite it to imply that an entryway can be a generic concept. Most developers implementing something like "sign in with your *.bsky.social account" will want it.

> The Authorization Server (PDS or [PDS Entryway](https://docs.bsky.app/docs/advanced-guides/entryway)) must implement a web interface for users to authenticate with the server, approve (or reject) Authorization Requests from clients, and manage active sessions. This is called the "Authorization Interface".

Author: Maybe we can do something like have a "How is OAuth used in AT Protocol?" or maybe just more clearly demark the section quoted above?

The authorization approval prompt should identify the client app and describe the scope of authorization that has been requested.

The amount of client metadata that should be displayed may depend on whether the client is "trusted" by the Authorization Server; see the ["Client"](#clients) and ["Security Considerations"](#security-considerations) sections. The full `client_id` URL should be displayed by default.

> 

Reviewer: Can we keep this for now (and discuss OIDC in another PR/Discussion) ?

Author: Oh, okay. Interesting. So yeah, I'd need to revert this change.

Author: @matthieusieben just to double check my understanding: this is silent SSO when the client is trusted or a confidential client, and there's no change in previously granted scopes, and no `prompt` parameter, correct?

- [OpenID Connect Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest)
- [Initiating User Registration via OpenID Connect 1.0](https://openid.net/specs/openid-connect-prompt-create-1_0.html)
- JSON Web Token (JWT) Profile for OAuth 2.0 Client Authentication and Authorization Grants, [RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523)
- OAuth 2.0 Demonstrating Proof of Possession (DPoP), [RFC 9449](https://datatracker.ietf.org/doc/html/rfc9449)
- OAuth 2.0 Protected Resource Metadata, [RFC 9728](https://datatracker.ietf.org/doc/html/rfc9728)
- OAuth 2.0 Authorization Server Metadata, [RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414)
- OAuth 2.0 Authorization Server Issuer Identification, [RFC 9207](https://datatracker.ietf.org/doc/html/rfc9207)

The following IANA Registries are referenced in this document:

- [OAuth parameter registry](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml)

The following other documents are referenced in this document:
- [HTTP CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS)
- [PDS Entryway](https://docs.bsky.app/docs/advanced-guides/entryway)

> - ["Apple Universal Links"](https://developer.apple.com/documentation/xcode/allowing-apps-and-websites-to-link-to-your-content)

Author: It would be good to include a link to NIST "P-256" here, which is referenced above, however, I could not find an exact link for this document.
